packages/angular_devkit/schematics/src/engine/task.ts

```diff
@@ -0,0 +1,83 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import { PriorityQueue } from '@angular-devkit/core';
+import { Observable } from 'rxjs/Observable';
+import { SchematicContext } from './interface';
+
+
+export interface TaskConfiguration<T = {}> {
+  name: string;
+  dependencies?: Array<TaskId>;
+  options?: T;
+}
+
+export interface TaskConfigurationGenerator<T = {}> {
+  toConfiguration(): TaskConfiguration<T>;
+}
+
+export type TaskExecutor<T = {}>
+  = (options: T | undefined, context: SchematicContext) => Observable<void>;
+
+export interface TaskExecutorFactory<T> {
+  readonly name: string;
+  create(options?: T): Promise<TaskExecutor>;
+}
+
+export interface TaskId {
+  readonly id: number;
+  readonly priority:  number;
+  readonly configuration: Readonly<TaskConfiguration>;
+  readonly context: SchematicContext;
+}
+
+function compareTasks(x: TaskId, y: TaskId) {
+  if (x.priority === y.priority) {
+    return 0;
+  }
+
+  return x.priority > y.priority ? 1 : -1;
+}
+
+export class TaskScheduler {
+  private _queue = new PriorityQueue<TaskId>(compareTasks);
```

Comment:
This seems like it could be simplified to `(x, y) => x.priority - y.priority`?

Reply:
I kind of liked that the return type of the compare function was `1 | 0 | -1` but that's an arbitrary reason to keep it.

Done.